src/protocol/CommitmentSyncer.sol

```diff
@@ -0,0 +1,109 @@
+    error UnauthorizedCheckpointTracker();
+
+    address private immutable _checkpointTracker;
+
+    mapping(uint64 chainId => uint64) private _latestHeight;
+    mapping(uint64 chainId => mapping(uint64 height => bytes32)) private _commitment;
+
+    /// @dev Reverts if the caller is not the `checkpointTracker`.
+    modifier onlyCheckpointTracker() {
+        _checkCheckpointTracker(msg.sender);
+        _;
+    }
+
+    /// @dev Sets the checkpoint tracker.
+    constructor(address checkpointTracker_) {
+        _checkpointTracker = checkpointTracker_;
+    }
+
+    /// @dev Contract that tracks commitment checkpoints.
+    function checkpointTracker() public view virtual returns (ICheckpointTracker) {
+        return ICheckpointTracker(_checkpointTracker);
+    }
+
+    /// @inheritdoc ICommitmentSyncer
+    function commitmentId(uint64 chainId, uint64 height, bytes32 commitment) public pure virtual returns (bytes32 id) {
+        return keccak256(abi.encodePacked(chainId, height, commitment));
+    }
+
+    /// @inheritdoc ICommitmentSyncer
+    function verifyCommitment(uint64 chainId, uint64 height, bytes32 commitment, bytes32 root, bytes[] calldata proof)
+        public
+        view
+        virtual
+        returns (bool valid, bytes32 id)
+    {
+        id = commitmentId(chainId, height, commitment);
+        return (LibTrieProof.verifyState(id.deriveSlot(), id, root, proof), id);
+    }
+
+    /// @inheritdoc ICommitmentSyncer
+    function commitmentAt(uint64 chainId, uint64 height) public view virtual returns (bytes32 commitment) {
+        return _commitment[chainId][height];
+    }
+
+    /// @inheritdoc ICommitmentSyncer
+    function latestCommitment(uint64 chainId) public view virtual returns (bytes32 commitment) {
+        return commitmentAt(chainId, latestHeight(chainId));
+    }
+
+    /// @inheritdoc ICommitmentSyncer
+    function latestHeight(uint64 chainId) public view virtual returns (uint64 height) {
+        return _latestHeight[chainId];
+    }
+
+    /// @inheritdoc ICommitmentSyncer
+    function syncCommitment(uint64 chainId, uint64 height, bytes32 commitment, bytes32 root, bytes[] calldata proof)
+        external
+        virtual
+        onlyCheckpointTracker
+    {
+        _syncCommitment(_checkCommitment(chainId, height, commitment, root, proof), chainId, height, commitment);
+    }
+
+    /// @dev Internal version of `syncCommitment` without access control.
+    /// Emits `CommitmentSynced` if the provided `height` is larger than `latestHeight` for `chainId`.
+    function _syncCommitment(bytes32 id, uint64 chainId, uint64 height, bytes32 commitment) internal virtual {
+        if (latestHeight(chainId) < height) {
+            _latestHeight[chainId] = height;
+            _commitment[chainId][height] = commitment;
+            // Invariant:
+            // assert(id == commitmentId(chainId, height, commitment));
+            id.signal();
```

Comment:
So now there are two SSTORE: `_commitment[chainId][height] = commitment` and `id.signal()`, can we just keep the latter?